### PR TITLE
Documented the overridden form options

### LINKED ISSUE #3412: Documenting overriden options
A lot overriden options are not yet documented. For information how to document these, refer to https://github.com/symfony/symfony-docs/issues/2361

# Missing overriden options

[**button**](http://symfony.com/doc/current/reference/forms/types/button.html)
 - [x] [`auto_initialize`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/ButtonType.php#L48)

[**checkbox**](http://symfony.com/doc/current/reference/forms/types/checkbox.html)
 - [x] [`empty_data`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/CheckboxType.php#L60)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/CheckboxType.php#L61)

[**choice**](http://symfony.com/doc/current/reference/forms/types/choice.html)
 - [x] [`empty_data`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/ChoiceType.php#L216)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/ChoiceType.php#L219)
 - [x] [`error_bubbling`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/ChoiceType.php#L218)

[**datetime**](http://symfony.com/doc/current/reference/forms/types/datetime.html)
 - [x] [`by_reference`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/DateTimeType.php#L221-223)
 - [x] [`error_bubbling`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/DateTimeType.php#L224)
 - [x] [`data_class`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/DateTimeType.php#L225-229)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/DateTimeType.php#L230)

[**date**](http://symfony.com/doc/current/reference/forms/types/date.html)
 - [x] [`data_class`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/DateType.php#L206-210)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/DateType.php#L211)

[**file**](http://symfony.com/doc/current/reference/forms/types/file.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/FileType.php#L53)
 - [x] [`data_class`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/FileType.php#L54)
 - [x] [`empty_data`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/FileType.php#L55)

[**hidden**](http://symfony.com/doc/current/reference/forms/types/hidden.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/HiddenType.php#L29)

[**integer**](http://symfony.com/doc/current/reference/forms/types/integer.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/IntegerType.php#L45)

[**money**](http://symfony.com/doc/current/reference/forms/types/money.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/MoneyType.php#L58)

[**number**](http://symfony.com/doc/current/reference/forms/types/number.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/NumberType.php#L43)

[**password**](http://symfony.com/doc/current/reference/forms/types/password.html)
 - [x] [`trim`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/PasswordType.php#L38)

[**percent**](http://symfony.com/doc/current/reference/forms/types/percent.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/PercentType.php#L37)

[**text**](http://symfony.com/doc/current/reference/forms/types/text.html)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/TextType.php#L25)

[**time**](http://symfony.com/doc/current/reference/forms/types/time.html)
 - [x] [`data_class`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/TimeType.php#L191-195)
 - [x] [`compound`](https://github.com/symfony/Form/blob/master/Extension/Core/Type/TimeType.php#L196)

[**timezone**](http://symfony.com/doc/current/reference/forms/types/timezone.html)
 - [x] overrides `choices` instead of `choice_list`

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | #3412 |

### CODE REVIEW

**Comment by WouterJ** [comment]
On the first question: Yes, you don't have to use includes. It's just to avoid duplication. If there still is a lot of copy/past, you can decide to use placeholders like in [`empty_value.rst.inc`](https://github.com/symfony/symfony-docs/blob/2.7/reference/forms/types/options/empty_data.rst.inc) or seperate it like `_error_bubbling_body.rst.inc` and `_error_bubbling_hint.rst.inc`.

For the side note, it already has an issue: https://github.com/symfony/symfony-docs/issues/5213

**Comment by javiereguiluz** [comment]
This PR is now ready to be reviewed. Thanks.

File: reference/forms/types/options/compound_type.rst.inc [hunk 1]
```diff
@@ -0,0 +1,8 @@
+compound
+~~~~~~~~
+
+**type**: ``boolean`` **default**: ``false``
+
+This option specifies if the type is compound. This is independent of whether
```
Comment (timglabisch) [inline reference/forms/types/options/compound_type.rst.inc]: this doesn't explain what "compound" means.

Comment (javiereguiluz) [inline reference/forms/types/options/compound_type.rst.inc]: You are right! I'm waiting for the general fix of this issue: https://github.com/symfony/symfony-docs/issues/5213

File: reference/forms/types/file.rst [hunk 2]
```diff
@@ -82,14 +105,8 @@ type:
 
 .. include:: /reference/forms/types/options/disabled.rst.inc
 
-.. include:: /reference/forms/types/options/empty_data.rst.inc
-    :end-before: DEFAULT_PLACEHOLDER
-
 The default value is ``null``.
```
Comment (WouterJ) [inline reference/forms/types/file.rst]: this can be removed too?

File: reference/forms/types/password.rst [hunk 3]
```diff
@@ -21,13 +24,25 @@ The ``password`` field renders an input password text box.
...
+Overridden Options
+------------------
+
+trim
+~~~~
+
+**default**: ``false``
+
+Unlike the rest of form types, the ``password`` type doesn't apply the
+phpfunction:`trim` function to the value submitted by the user. This ensures that
```
Comment (WouterJ) [inline reference/forms/types/password.rst]: missing `:` before `phpfunction`

File: reference/forms/types/date.rst [hunk 4]
```diff
@@ -30,6 +30,8 @@ day and year) or three select boxes (see the `widget`_ option).
 |                      | - `years`_                                                                  |
 +----------------------+-----------------------------------------------------------------------------+
 | Overridden Options   | - `by_reference`_                                                           |
```
Comment (xabbuh) [inline reference/forms/types/date.rst]: options imho should be lowercased (to be consistent with "Inherited options").

File: reference/forms/types/options/data_class_date.rst.inc [hunk 5]
```diff
@@ -0,0 +1,8 @@
+data_class
+~~~~~~~~~~
+
+**default**: ``null``
```
Comment (xabbuh) [inline reference/forms/types/options/data_class_date.rst.inc]: Missing `type` here:

```rst
**type**: ``string`` **default**: ``null``
```

**Comment by javiereguiluz** [comment]
I've updated the description of the `compound` option. This PR is now ready for the final review. Thanks.

File: reference/forms/types/datetime.rst [hunk 6]
```diff
@@ -15,6 +15,11 @@ the data can be a ``DateTime`` object, a string, a timestamp or an array.
 +----------------------+-----------------------------------------------------------------------------+
 | Rendered as          | single text box or three select fields                                      |
 +----------------------+-----------------------------------------------------------------------------+
+| Overridden options   | - `by_reference`_                                                           |
+|                      | - `compound`_                                                               |
+|                      | - `data_class`_                                                             |
+|                      | - `error_bubbling`_                                                         |
++----------------------+-----------------------------------------------------------------------------+
```
Comment (xabbuh) [inline reference/forms/types/datetime.rst]: This block should be moved below the "Options" block.

Comment (xabbuh) [inline reference/forms/types/datetime.rst]: and the same applies to the actual descriptions below

File: reference/forms/types/file.rst [hunk 7]
```diff
@@ -24,6 +27,26 @@ The ``file`` type represents a file input in your form.
...
 
+Overridden Options
+------------------
+
+.. include:: /reference/forms/types/options/compound_type.rst.inc
+
+data_class
+~~~~~~~~~~
+
+**default**: :class:`Symfony\\Component\\HttpFoundation\\File\\File`
```
Comment (xabbuh) [inline reference/forms/types/file.rst]: missing type

File: reference/forms/types/file.rst [hunk 8]
```diff
@@ -24,6 +27,26 @@ The ``file`` type represents a file input in your form.
...
+
+This option sets the appropriate file-realted data mapper to be used by the type.
+
+empty_data
+~~~~~~~~~~
+
+**default**: ``null``
```
Comment (xabbuh) [inline reference/forms/types/file.rst]: same here

File: reference/forms/types/password.rst [hunk 9]
```diff
@@ -9,6 +9,9 @@ The ``password`` field renders an input password text box.
 +-------------+------------------------------------------------------------------------+
 | Rendered as | ``input`` ``password`` field                                           |
 +-------------+------------------------------------------------------------------------+
+| Overridden  | - `trim`_                                                              |
+| options     |                                                                        |
++-------------+------------------------------------------------------------------------+
```
Comment (xabbuh) [inline reference/forms/types/password.rst]: should be moved below "Options" (same for their descriptions)

File: reference/forms/types/password.rst [hunk 10]
```diff
@@ -21,13 +24,25 @@ The ``password`` field renders an input password text box.
...
 | Class       | :class:`Symfony\\Component\\Form\\Extension\\Core\\Type\\PasswordType` |
 +-------------+------------------------------------------------------------------------+
 
+Overridden Options
+------------------
+
+trim
+~~~~
+
+**default**: ``false``
```
Comment (xabbuh) [inline reference/forms/types/password.rst]: missing "type"

**Comment by xabbuh** [comment]
@javiereguiluz I left you some last minor comments. After that, this looks ready to be merged.

**Comment by WouterJ** [comment]
@xabbuh didn't we remove types from the overriden options? (as it's more about documenting the new value and why than actually documenting the option)

**Comment by xabbuh** [comment]
@WouterJ At least, we are not consistent with this rule then: http://symfony.com/doc/current/reference/forms/types/entity.html#choice-list

By the way, does it really make sense to leave out the type? I would then always have to look the type up in the parent type. That's not really comfortable, is it?

**Comment by javiereguiluz** [comment]
@xabbuh I've fixed all the issues that you found and I've rebased the PR. I guess this is now ready to be merged. Thanks!

File: reference/forms/types/entity.rst [hunk 11]
```diff
@@ -19,8 +19,8 @@ objects from the database.
 |             | - `property`_                                                    |
 |             | - `query_builder`_                                               |
 +-------------+------------------------------------------------------------------+
-| Overriden   | - `choice_list`_                                                 |
-| options     | - `choices`_                                                     |
+| Overridden  | - `choices`_                                                     |
+| options     | - `choice_list`_                                                 |
```
Comment (xabbuh) [inline reference/forms/types/entity.rst]: This should be reverted (see #5343).

File: reference/forms/types/file.rst [hunk 12]
```diff
@@ -24,6 +27,26 @@ The ``file`` type represents a file input in your form.
...
+------------------
+
+.. include:: /reference/forms/types/options/compound_type.rst.inc
+
+data_class
+~~~~~~~~~~
+
+**type**: ``string`` **default**: :class:`Symfony\\Component\\HttpFoundation\\File\\File`
+
+This option sets the appropriate file-realted data mapper to be used by the type.
```
Comment (xabbuh) [inline reference/forms/types/file.rst]: related

File: reference/forms/types/entity.rst (hunk 11 above)
Comment (javiereguiluz) [inline reference/forms/types/entity.rst]: Done.

File: reference/forms/types/file.rst (hunk 12 above)
Comment (javiereguiluz) [inline reference/forms/types/file.rst]: Good catch! Thanks.

**Comment by xabbuh** [comment]
:+1: Looks great now!

**Comment by weaverryan** [comment]
Wow, BIG work - that's incredible. Thanks Javier!

So now, as long as code changes show up as docs issues/PR's, we can stay in sync permanently.

Minor reordering tweaks at sha: f28ef78de4b0e7028056ca2d7f9c453f652345d0. There is some inconsistency with the order of the sections, but I think that's ok - sometimes overridden seems more important than inherited and vice-versa. And really, in a perfect world, I'm not sure if we need these different sections at all - does a user care of it's inherited? And I might also want the order of the options to be non-alphabetical, and instead to be based on how common an option is. Just thinking aloud :)

Cheers!

**Pull request merged**